### PR TITLE
Python: New Arrow Fathead (#669)

### LINKED ISSUE #669: Python: Create New arrow Reference Fathead
# Create a new *arrow* Fathead Instant Answer
Help us make DuckDuckGo the best search engine for programmers!

### What do I need to know?
You'll need to know how to code in **Perl**, **Python**, **Ruby**, or **JavaScript**.
fathead languages

### What am I doing?
You will write a script that scrapes or downloads the data source below, and generates an **output.txt** file containing the parsed documentation. You can learn more about Fatheads and the `output.txt` syntax [**here**](https://docs.duckduckhack.com/resources/fathead-overview.html).

**Data source**: https://arrow.readthedocs.io/en/latest/

**Instant Answer Page**: *to be created by developer*

### What is the Goal?
As part of our [Programming Mission](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53), we're aiming to reach 100% Instant Answer (IA) coverage for searches related to programming languages by creating new Instant Answers, and improving existing ones.

Here are some Fathead examples:

- Ruby Docs
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/ruby) | [Example Query](https://duckduckgo.com/?q=array+bsearch&ia=about)
- MDN CSS
    - [Code](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/mdn_css) | [Example Query](https://duckduckgo.com/?q=css+background-position&ia=about)

  css background-position search

[See more related Instant Answers](https://duck.co/ia?repo=fathead)

## Get Started
- [x] 1) Claim this issue by commenting below
- [x] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [x] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [x] 4) Create a new Instant Answer Page: https://duck.co/ia/new_ia (then let us know, here!)
- [ ] 5) Create the Fathead
- [ ] 6) Create a Pull Request
- [ ] 7) Ping @pjhampton for a review

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

### PR DESCRIPTION
**Pull request by danidee10**
## Description of new Instant Answer, or changes

This PR adds a new fathead for [Arrow](http://crsmithdev.com/arrow/).
Arrow is a Python library that offers a sensible, human-friendly approach to creating, manipulating, formatting and converting dates, times, and timestamps

## Related Issues and Discussions

Fixes #669 

## People to notify

@pjhampton @moollaza 

## Testing & Review
To be completed by Language Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [x] Title follows correct format (Specifies Instant Answer + Purpose)
- [x] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [x] Instant Answer page is correctly filled out and contains:
    - One topic for the Search Space Language (Java, Python, Scala, Ruby, etc.)
    - One topic from: Reference, Help, Libraries, Tools
        - Documentation Fatheads are considered "Reference"
    - Description, Data source, and 2+ example queries
    - Perl Module (e.g. "DDG::Fathead::PerlDoc" -- we only need a name, not an actual file)
    - Source Name (for "More at <source_name>" link)
    - Source Domain (must contain http:// or https:// -- can be the same as Data Source)
    - Source Info (used as Subtitle for each Article -- usually matches the IA Name)
    - 'Skip Abstract' is checked off
    - Source ID (ping @moollaza to assign one, once Fathead is ready for Beta deploy)

**Code**
- [x] Uniformly indented, well commented
- [ ] Fetch.sh and Parse.sh run without errors
- [ ] Output contains no blank lines, or multi-line entries
- [ ] Fathead Tests are passing (run `$ duckpan test <fathead_id>`)
    - Tester should report any failures

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/arrow_docs

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
@danidee10 this is looking great! It's now live on Beta: https://beta.duckduckgo.com/?q=arrow.arrow.Arrow.clone&ia=about

It looks like there's one minor issue, you have an empty paragraph (`<p></p>`) creating extra whitespace.

File: lib/fathead/arrow_docs/parse.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+"""Parse Arrow docs."""
+
+from bs4 import BeautifulSoup
+import codecs
+import re
+
+INFILE = 'download/docs.html'
+OUTFILE = 'output.txt'
+
+abstract_fmt = '<section class="prog__container"><p>{}{}</p></section>'
```
Comment (moollaza) [inline lib/fathead/arrow_docs/parse.py]: @danidee10 the `<p>` tag should not include the code as well. The `<p>` and `<pre><code>` should be at the same level:

```html
<section>
  <p>$text</p>
  <pre>
    <code>$code</code>
  </pre>
</section>
```

When there's no code, there should just be `<section><p>$text</p></section>` 👍

Comment (danidee10) [inline lib/fathead/arrow_docs/parse.py]: @moollaza I just pushed a commit that addresses this issue.

**Comment by moollaza** [comment]
@danidee10 Thanks! This LGTM :+1:

**Comment by moollaza** [comment]
@danidee10 you can remove the empty `trigger_words.txt` and `redirects.txt` 

I would suggest to make this easier to trigger, you create a redirect entry for each article where you replace `arrow.arrow.Arrow` with just `arrow `.

i.e. `arrow clone` -> `arrow.arrow.Arrow.clone`

That way a search for "arrow clone" would show the article for `arrow.arrow.Arrow.clone`

**Comment by daxtheduck** [comment]
@danidee10 this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Pull request merged**